### PR TITLE
improve preprocessing performance

### PR DESCRIPTION
**Pull request by StefanVukovic99**
This should improve performance on the translator benchmark by ~30% and get rid of some of the complicated stuff with option spaces.

### CODE REVIEW

**Comment by jamesmaa** [comment]
> This should improve performance on the translator benchmark by ~30% and get rid of some of the complicated stuff with option spaces.

How are you measuring 30%?

File: ext/js/language/translator.js [hunk 1]
```diff
@@ -447,16 +436,10 @@ export class Translator {
         const {language} = options;
         const info = this._textProcessors.get(language);
```
Comment (jamesmaa) [inline ext/js/language/translator.js:437]: Why is this variable named `info`?

File: ext/js/language/translator.js [hunk 2]
```diff
@@ -497,36 +469,59 @@ export class Translator {
...
+        for (const {id, textProcessor: {process, options}} of textProcessors) {
+            /** @type {Set<string>} */
+            const newVariants = new Set();
+            for (const variant of variants) {
+                for (const option of options) {
+                    const processed = this._getProcessedText(textCache, variant, id, option, process);
+                    newVariants.add(processed);
+                }
             }
+            variants = newVariants;
```
Comment (jamesmaa) [inline ext/js/language/translator.js:493]: doesn't `variants` get overshadowed every time this for-loop runs?

**Review by jamesmaa — COMMENTED** [review]
Could you write in the PR description what high level changes you've made? I think it'll help frame the PR a lot better. It does look simpler but reasoning about the logic without a high level understanding of what's intended is kind of annoying

**Comment by StefanVukovic99** [comment]
> How are you measuring 30%?

https://github.com/themoeway/yomitan/pull/973/checks?check_run_id=25239991743

i was rounding it up to 1s -> 0.7s

File: ext/js/language/translator.js (hunk 2 above)
Comment (StefanVukovic99) [reply ext/js/language/translator.js:493]: It's hard to explain these things, but essentially we are starting with a set that contains just the raw word, and then for each cycle of the loop we go one step deeper on the tree, replacing that word with every variant given by the current preprocessor.

To borrow the example from the docs for english we go

1. `['reaD']`
2. `['reaD', 'read']` (variants for `decapitalize`)
3. `['reaD', 'ReaD', 'read', 'Read'] (variants for each of the last two for `capitalizeFirstLetter`)

You can see how they form a tree, but we only need to remember the current layer.

**Comment by StefanVukovic99** [comment]
> high level changes you've made?

Other than the comment about the tree, I'd say that instead of generating each variant, checking if it has been used before and then deinflecting it, we now generate a set of all possible variants upfront and deinflect each.

ChatGPT gave me some ascii
old way was like 
```
1  5   9   13
|  |   |    |
2  6   10   14
|  |   |    |
3  7   11   15
|  |   |    |
4  8   12   16
```
new way is like
```
       1
      / \
     2   3
    /|   |\
   4 5   6 7
```
lot less operations.

**Review by jamesmaa — APPROVED** [review]
This implementation is definitely simpler. Glad it's more performant too

**Pull request merged**